Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -113,141 +114,210 @@ end Ideal
+  r M N := ∀ i j, c (M i j) (N i j)
+  -- note: kept `fun` to distinguish `RingCon`'s binders from `r`'s binders.
+  iseqv.refl _ := fun _ _ ↦ c.refl _
+  iseqv.symm h := fun _ _ ↦ c.symm <| h _ _
+  iseqv.trans h₁ h₂ := fun _ _ ↦ c.trans (h₁ _ _) (h₂ _ _)
+  add' h₁ h₂ := fun _ _ ↦ c.add (h₁ _ _) (h₂ _ _)
+  mul' h₁ h₂ := fun _ _ ↦ c.finset_sum _ fun _ _ => c.mul (h₁ _ _) (h₂ _ _)
+
+@[simp]
+theorem matrix_apply {c : RingCon R} {M N : Matrix n n R} :
+    c.matrix n M N ↔ ∀ i j, c (M i j) (N i j) :=
+  Iff.rfl
+
+@[simp]
+theorem matrix_apply_stdBasisMatrix [DecidableEq n] {c : RingCon R} {i j : n} {x y : R} :
+    c.matrix n (Matrix.stdBasisMatrix i j x) (Matrix.stdBasisMatrix i j y) ↔ c x y := by
+  refine ⟨fun h ↦ by simpa using h i j, fun h i' j' ↦ ?_⟩
+  obtain hi | rfl := ne_or_eq i i'
+  · simpa [hi] using c.refl 0
+  obtain hj | rfl := ne_or_eq j j'
+  · simpa [hj] using c.refl _
+  simpa using h
+
+theorem matrix_monotone : Monotone (matrix (R := R) n) :=
+  fun _ _ hc _ _ h _ _ ↦ hc (h _ _)
+
+theorem matrix_injective [Nonempty n] : Function.Injective (matrix (R := R) n) :=
+  fun I J eq ↦ RingCon.ext fun r s ↦ by
+    have := congr_fun (DFunLike.congr_fun eq (Matrix.of fun _ _ ↦ r)) (Matrix.of fun _ _ ↦ s)
+    simpa using this
+
+theorem matrix_strictMono_of_nonempty [Nonempty n] :
+    StrictMono (matrix (R := R) n) :=
+  matrix_monotone n |>.strictMono_of_injective <| matrix_injective _
+
+@[simp]
+theorem matrix_bot : (⊥ : RingCon R).matrix n = ⊥ :=
+  eq_bot_iff.2 fun _ _ h ↦ Matrix.ext h
+
+@[simp]
+theorem matrix_top : (⊤ : RingCon R).matrix n = ⊤ :=
+  eq_top_iff.2 fun _ _ _ _ _ ↦ trivial
+
+open Matrix
+
+variable {n}
+
+/-- The congruence relation induced by `c` on `stdBasisMatrix i j`. -/
+def ofMatrix [DecidableEq n] (c : RingCon (Matrix n n R)) : RingCon R where
+  r x y := ∀ i j, c (stdBasisMatrix i j x) (stdBasisMatrix i j y)
+  iseqv.refl _ := fun _ _ ↦ c.refl _
+  iseqv.symm h := fun _ _ ↦ c.symm <| h _ _
+  iseqv.trans h₁ h₂ := fun _ _ ↦ c.trans (h₁ _ _) (h₂ _ _)
+  add' h₁ h₂ := fun _ _ ↦ by simpa [stdBasisMatrix_add] using c.add (h₁ _ _) (h₂ _ _)
+  mul' h₁ h₂ := fun i j ↦ by simpa using c.mul (h₁ i i) (h₂ i j)
+
+@[simp]
+theorem ofMatrix_rel [DecidableEq n] {c : RingCon (Matrix n n R)} {x y : R} :
+    ofMatrix c x y ↔ ∀ i j, c (stdBasisMatrix i j x) (stdBasisMatrix i j y) :=
+  Iff.rfl
+
+@[simp] theorem ofMatrix_matrix [DecidableEq n] [Nonempty n] (c : RingCon R) :
+    ofMatrix (matrix n c) = c := by
+  ext x y
+  classical
+  constructor
+  · intro h
+    inhabit n
+    simpa using h default default default default
+  · intro h i j
+    rwa [matrix_apply_stdBasisMatrix]
+
+end NonUnitalNonAssocSemiring
+
+section NonAssocSemiring
+variable [NonAssocSemiring R] [Fintype n]
+open Matrix
+
+/-- Note that this does not apply to a non-unital ring, with counterexample where the elementwise
+congruence relation `!![⊤,⊤;⊤,c']` is a ring congruence over `Matrix (Fin 2) (Fin 2) R` for any
+ring congruence `c'`. -/
```

Comment:
@alreadydone, is this an accurate lean-ification of your example?

Reply:
My example depends on the fact that `2ℤ * 2ℤ ⊆ 4ℤ`. Certainly if R is a unital ring then `!![⊤,⊤;⊤,c']` being a RingCon implies that `⊤ = c'`.

Reply:
I think I got a little confused about whether `R` was the ring or the relation. Is the following better?
```suggestion
/-- Note that this does not apply to a non-unital ring, with counterexample where the elementwise
congruence relation `!![2ℤ,2ℤ;2ℤ,4ℤ]` is a ring congruence over `Matrix (Fin 2) (Fin 2) ℤ`. -/
```

Reply:
It's a ring congruence over `Matrix (Fin 2) (Fin 2) 2ℤ`, otherwise looks good!

Reply:
So is this
```suggestion
/-- Note that this does not apply to a non-unital ring, with counterexample where the elementwise
congruence relation `!![⊥,⊥;⊥, (· ≡ · [PMOD 4])]` is a ring congruence over `Matrix (Fin 2) (Fin 2) 2ℤ`. -/
```
correct?

Reply:
⊥ should be ⊤, otherwise LGTM